File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/httpdir/Catalog.java
/**
 * @file
 * @brief Catalog interface
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.httpdir;

import android.content.Context;

import java.io.IOException;

import app.zxtune.fs.http.MultisourceHttpProvider;

public abstract class Catalog {

  public interface DirVisitor {
    void acceptDir(String name, String description);

    void acceptFile(String name, String description, String size);
  }

  /**
   * @param path    resource location
   * @param visitor result receiver
   */
  public abstract void parseDir(Path path, DirVisitor visitor) throws IOException;

  public static CachingCatalog create(Context ctx, MultisourceHttpProvider http, String id) {
    final RemoteCatalog remote = new RemoteCatalog(http);
    return create(ctx, remote, id);
  }

  public static CachingCatalog create(Context ctx, RemoteCatalog remote, String id) {
    return new CachingCatalog(ctx, remote, id);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/httpdir/Catalog.kt
package app.zxtune.fs.httpdir

import android.content.Context
import app.zxtune.fs.http.MultisourceHttpProvider
import java.io.IOException

interface Catalog {

    interface DirVisitor {
        fun acceptDir(name: String, description: String)

        fun acceptFile(name: String, description: String, size: String)
    }

    /**
     * @param path    resource location
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun parseDir(path: Path, visitor: DirVisitor)

    companion object {
        @JvmStatic
        fun create(ctx: Context, http: MultisourceHttpProvider, id: String): Catalog {
            val remote = RemoteCatalog(http)
            return create(ctx, remote, id)
        }

        @JvmStatic
        fun create(ctx: Context, remote: RemoteCatalog, id: String): Catalog =
            CachingCatalog(ctx, remote, id)
    }
}
